{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run coloc vs. liver and skin traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading packages...\n",
      "Loading input files...\n",
      "Loading functions...\n",
      "Loading input data...\n",
      "Loading metadata...\n"
     ]
    }
   ],
   "source": [
    "invisible(suppressWarnings(file.link(\"/frazer01/home//matteo/notebooks/eqtls_deconvolution_gtex//coloc_analysis.ipynb\"     , \"analysis/coloc_analysis.ipynb\"    )))\n",
    "invisible(suppressWarnings(file.link(\"/frazer01/home//matteo/notebooks/eqtls_deconvolution_gtex//cardiac_qtls_run_coloc.R\" , \"analysis/cardiac_qtls_run_coloc.R\")))\n",
    "invisible(suppressWarnings(file.link(\"/frazer01/home//matteo/notebooks/eqtls_deconvolution_gtex//cardiac_qtls_run_coloc.sh\", \"analysis/cardiac_qtls_run_coloc.sh\")))\n",
    "\n",
    "source(\"analysis/cardiac_qtls_packages.R\"      )\n",
    "source(\"analysis/cardiac_qtls_input_files.R\"   )\n",
    "source(\"analysis/cardiac_qtls_functions.R\"     )\n",
    "source(\"analysis/cardiac_qtls_input_data.R\"    )\n",
    "source(\"analysis/cardiac_qtls_load_metadata.R\" )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(\"private\", \"coloc\",          sep = \"/\"), showWarnings = FALSE)\n",
    "\n",
    "mains   = c(\"liver\", \"skin\")\n",
    "invisible(lapply(mains, function(main){dir.create(paste(\"private\", \"coloc\", main, sep = \"/\"), showWarnings = FALSE)}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues         = c(\"liver_original\", \"liver_cells\", \"liver2_cells\", \"liver3_cells\", \"skin_original\", \"skin_cells\", \"skin2_cells\"  )\n",
    "tissue2name     = data.frame(tissue = tissues, name = c(\"Liver (original)\", \"Liver (mouse cell populations)\", \"Liver (human cell populations)\", \"Liver (collapsed resolution)\", \"Skin (original)\", \"Skin (mouse cell populations)\" , \"Skin (collapsed resolution)\" ), color = c(\"#FF7256\", \"#CD3700\", \"#800000\", \"#ff0000\", \"#87CEFA\", \"#000080\", \"#8080ff\"), y = c(7,6,5,3,4,2,1), main = c(rep(\"liver\", 4), rep(\"skin\", 3)))\n",
    "qtl_list        = lapply(tissues , function(x){read.table(paste(\"qtls/\", x, \"/analysis/egenes.\", x, \".txt\", sep = \"\"), header = TRUE)})\n",
    "names(qtl_list) = tissues\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_egenes = function(qtl_list, tissue)\n",
    "{\n",
    "    x      = qtl_list[[tissue]]\n",
    "    egenes = x[x$egene == TRUE, \"gene_id\"]\n",
    "    \n",
    "    return(data.frame(gene_id = egenes, tissue = tissue))\n",
    "}\n",
    "egenes = do.call(\"rbind\", lapply(tissues, function(tissue){get_egenes(qtl_list, tissue)}))\n",
    "egenes = merge(egenes, tissue2name)\n",
    "egenes = unique(egenes[,c(\"gene_id\", \"main\")])\n",
    "\n",
    "write.table(egenes, paste(\"private\", \"coloc\", \"egenes.txt\", sep = \"/\"), sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "trait2test = list(liver = read_gwas_table(\"/home/mdonovan/gtex_deconvolution/tables/liver_gwas.txt\"), skin = read_gwas_table(\"/home/mdonovan/gtex_deconvolution/tables/skin_gwas.txt\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ids  = sort(unique(egenes$gene_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'qsub -l week -pe smp 1 -t 1-16920 -tc 500 -o /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/log/coloc.out -e /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/log/coloc.err /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/analysis/cardiac_qtls_run_coloc.sh'"
      ],
      "text/latex": [
       "'qsub -l week -pe smp 1 -t 1-16920 -tc 500 -o /frazer01/projects/GTEx\\_v7/analysis/eqtls\\_deconvolution/log/coloc.out -e /frazer01/projects/GTEx\\_v7/analysis/eqtls\\_deconvolution/log/coloc.err /frazer01/projects/GTEx\\_v7/analysis/eqtls\\_deconvolution/analysis/cardiac\\_qtls\\_run\\_coloc.sh'"
      ],
      "text/markdown": [
       "'qsub -l week -pe smp 1 -t 1-16920 -tc 500 -o /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/log/coloc.out -e /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/log/coloc.err /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/analysis/cardiac_qtls_run_coloc.sh'"
      ],
      "text/plain": [
       "[1] \"qsub -l week -pe smp 1 -t 1-16920 -tc 500 -o /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/log/coloc.out -e /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/log/coloc.err /frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/analysis/cardiac_qtls_run_coloc.sh\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running qsub\n"
     ]
    }
   ],
   "source": [
    "run_qsub_coloc = function(qtl_folder, n_genes, run_qsub = FALSE, queue = \"week\", tc = 500)\n",
    "{\n",
    "    err_file = paste(qtl_folder, \"log\", \"coloc.out\", sep = \"/\")\n",
    "    out_file = paste(qtl_folder, \"log\", \"coloc.err\", sep = \"/\")\n",
    "    \n",
    "    suppressWarnings(file.remove(err_file))\n",
    "    suppressWarnings(file.remove(out_file))\n",
    "\n",
    "\n",
    "    qsub_options = paste(\"-l\", queue, \n",
    "                         \"-pe\", \"smp 1\", \n",
    "                         \"-t\", paste(1, n_genes, sep = \"-\"), \n",
    "                         \"-tc\", tc,\n",
    "                         \"-o\", err_file,\n",
    "                         \"-e\", out_file\n",
    "                        )\n",
    "\n",
    "    command = paste(\"qsub\", qsub_options, paste(qtl_folder, \"analysis/cardiac_qtls_run_coloc.sh\", sep = \"/\"))\n",
    "\n",
    "    if (run_qsub == TRUE ){message(\"Running qsub\")}\n",
    "    if (run_qsub == TRUE ){system(command)}\n",
    "    if (run_qsub == FALSE){return(command)}\n",
    "}\n",
    "\n",
    "run_qsub_coloc(getwd(), length(gene_ids), run_qsub = FALSE, queue = \"week\", tc = 500)\n",
    "run_qsub_coloc(getwd(), length(gene_ids), run_qsub = TRUE , queue = \"week\", tc = 500)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (main, taskid) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    suppressMessages(source(\"analysis/cardiac_qtls_packages.R\"))</span>\n",
       "<span style=white-space:pre-wrap>    suppressMessages(source(\"analysis/cardiac_qtls_input_files.R\"))</span>\n",
       "<span style=white-space:pre-wrap>    suppressMessages(source(\"analysis/cardiac_qtls_functions.R\"))</span>\n",
       "<span style=white-space:pre-wrap>    suppressMessages(source(\"analysis/cardiac_qtls_input_data.R\"))</span>\n",
       "<span style=white-space:pre-wrap>    suppressMessages(source(\"analysis/cardiac_qtls_load_metadata.R\"))</span>\n",
       "<span style=white-space:pre-wrap>    gene2main = read.table(paste(\"private\", \"coloc\", \"egenes.txt\", </span>\n",
       "<span style=white-space:pre-wrap>        sep = \"/\"), header = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>    gene_ids = sort(unique(gene2main$gene_id))</span>\n",
       "<span style=white-space:pre-wrap>    gene_id = gene_ids[[taskid]]</span>\n",
       "<span style=white-space:pre-wrap>    chrom = geneinfo[geneinfo$gene_id == gene_id, \"chrom\"]</span>\n",
       "<span style=white-space:pre-wrap>    from = geneinfo[geneinfo$gene_id == gene_id, \"start\"] - qtl_distance</span>\n",
       "<span style=white-space:pre-wrap>    to = geneinfo[geneinfo$gene_id == gene_id, \"start\"] + qtl_distance</span>\n",
       "<span style=white-space:pre-wrap>    if (file.exists(paste(\"/frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/qtls/\", </span>\n",
       "<span style=white-space:pre-wrap>        main, \"_original/input/phenotypes/\", gene_id, \".txt\", </span>\n",
       "<span style=white-space:pre-wrap>        sep = \"\")) == TRUE) {</span>\n",
       "<span style=white-space:pre-wrap>        traits = read_gwas_table(paste(\"/home/mdonovan/gtex_deconvolution/tables/\", </span>\n",
       "<span style=white-space:pre-wrap>            main, \"_gwas.txt\", sep = \"\"))</span>\n",
       "<span style=white-space:pre-wrap>        pv = get_ukbb_data(gene_id, main, chrom, from, to, traits, </span>\n",
       "<span style=white-space:pre-wrap>            \"PV\")</span>\n",
       "<span style=white-space:pre-wrap>        trait2coloc = lapply(traits, function(trait) {</span>\n",
       "<span style=white-space:pre-wrap>            trait2df(trait, pv)</span>\n",
       "<span style=white-space:pre-wrap>        })</span>\n",
       "<span style=white-space:pre-wrap>        qtl2coloc = lapply(tissue2name[tissue2name$main == main, </span>\n",
       "<span style=white-space:pre-wrap>            \"tissue\"], function(tissue) {</span>\n",
       "<span style=white-space:pre-wrap>            qtl2df(gene_id, tissue)</span>\n",
       "<span style=white-space:pre-wrap>        })</span>\n",
       "<span style=white-space:pre-wrap>        names(trait2coloc) = traits</span>\n",
       "<span style=white-space:pre-wrap>        names(qtl2coloc) = tissue2name[tissue2name$main == main, </span>\n",
       "<span style=white-space:pre-wrap>            \"tissue\"]</span>\n",
       "<span style=white-space:pre-wrap>        to_coloc = c(qtl2coloc, trait2coloc)</span>\n",
       "<span style=white-space:pre-wrap>        to_test = expand.grid(qtl = names(qtl2coloc), trait = names(trait2coloc), </span>\n",
       "<span style=white-space:pre-wrap>            stringsAsFactors = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>        run_coloc_all_interactions(gene_id, to_coloc, to_test, </span>\n",
       "<span style=white-space:pre-wrap>            main)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (main, taskid) \n",
       "\\{\n",
       "    suppressMessages(source(\"analysis/cardiac\\_qtls\\_packages.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac\\_qtls\\_input\\_files.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac\\_qtls\\_functions.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac\\_qtls\\_input\\_data.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac\\_qtls\\_load\\_metadata.R\"))\n",
       "    gene2main = read.table(paste(\"private\", \"coloc\", \"egenes.txt\", \n",
       "        sep = \"/\"), header = TRUE)\n",
       "    gene\\_ids = sort(unique(gene2main\\$gene\\_id))\n",
       "    gene\\_id = gene\\_ids{[}{[}taskid{]}{]}\n",
       "    chrom = geneinfo{[}geneinfo\\$gene\\_id == gene\\_id, \"chrom\"{]}\n",
       "    from = geneinfo{[}geneinfo\\$gene\\_id == gene\\_id, \"start\"{]} - qtl\\_distance\n",
       "    to = geneinfo{[}geneinfo\\$gene\\_id == gene\\_id, \"start\"{]} + qtl\\_distance\n",
       "    if (file.exists(paste(\"/frazer01/projects/GTEx\\_v7/analysis/eqtls\\_deconvolution/qtls/\", \n",
       "        main, \"\\_original/input/phenotypes/\", gene\\_id, \".txt\", \n",
       "        sep = \"\")) == TRUE) \\{\n",
       "        traits = read\\_gwas\\_table(paste(\"/home/mdonovan/gtex\\_deconvolution/tables/\", \n",
       "            main, \"\\_gwas.txt\", sep = \"\"))\n",
       "        pv = get\\_ukbb\\_data(gene\\_id, main, chrom, from, to, traits, \n",
       "            \"PV\")\n",
       "        trait2coloc = lapply(traits, function(trait) \\{\n",
       "            trait2df(trait, pv)\n",
       "        \\})\n",
       "        qtl2coloc = lapply(tissue2name{[}tissue2name\\$main == main, \n",
       "            \"tissue\"{]}, function(tissue) \\{\n",
       "            qtl2df(gene\\_id, tissue)\n",
       "        \\})\n",
       "        names(trait2coloc) = traits\n",
       "        names(qtl2coloc) = tissue2name{[}tissue2name\\$main == main, \n",
       "            \"tissue\"{]}\n",
       "        to\\_coloc = c(qtl2coloc, trait2coloc)\n",
       "        to\\_test = expand.grid(qtl = names(qtl2coloc), trait = names(trait2coloc), \n",
       "            stringsAsFactors = FALSE)\n",
       "        run\\_coloc\\_all\\_interactions(gene\\_id, to\\_coloc, to\\_test, \n",
       "            main)\n",
       "    \\}\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (main, taskid) \n",
       "{\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_packages.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_input_files.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_functions.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_input_data.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_load_metadata.R\"))\n",
       "    gene2main = read.table(paste(\"private\", \"coloc\", \"egenes.txt\", \n",
       "        sep = \"/\"), header = TRUE)\n",
       "    gene_ids = sort(unique(gene2main$gene_id))\n",
       "    gene_id = gene_ids[[taskid]]\n",
       "    chrom = geneinfo[geneinfo$gene_id == gene_id, \"chrom\"]\n",
       "    from = geneinfo[geneinfo$gene_id == gene_id, \"start\"] - qtl_distance\n",
       "    to = geneinfo[geneinfo$gene_id == gene_id, \"start\"] + qtl_distance\n",
       "    if (file.exists(paste(\"/frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/qtls/\", \n",
       "        main, \"_original/input/phenotypes/\", gene_id, \".txt\", \n",
       "        sep = \"\")) == TRUE) {\n",
       "        traits = read_gwas_table(paste(\"/home/mdonovan/gtex_deconvolution/tables/\", \n",
       "            main, \"_gwas.txt\", sep = \"\"))\n",
       "        pv = get_ukbb_data(gene_id, main, chrom, from, to, traits, \n",
       "            \"PV\")\n",
       "        trait2coloc = lapply(traits, function(trait) {\n",
       "            trait2df(trait, pv)\n",
       "        })\n",
       "        qtl2coloc = lapply(tissue2name[tissue2name$main == main, \n",
       "            \"tissue\"], function(tissue) {\n",
       "            qtl2df(gene_id, tissue)\n",
       "        })\n",
       "        names(trait2coloc) = traits\n",
       "        names(qtl2coloc) = tissue2name[tissue2name$main == main, \n",
       "            \"tissue\"]\n",
       "        to_coloc = c(qtl2coloc, trait2coloc)\n",
       "        to_test = expand.grid(qtl = names(qtl2coloc), trait = names(trait2coloc), \n",
       "            stringsAsFactors = FALSE)\n",
       "        run_coloc_all_interactions(gene_id, to_coloc, to_test, \n",
       "            main)\n",
       "    }\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (main, taskid) \n",
       "{\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_packages.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_input_files.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_functions.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_input_data.R\"))\n",
       "    suppressMessages(source(\"analysis/cardiac_qtls_load_metadata.R\"))\n",
       "    gene2main = read.table(paste(\"private\", \"coloc\", \"egenes.txt\", \n",
       "        sep = \"/\"), header = TRUE)\n",
       "    gene_ids = sort(unique(gene2main$gene_id))\n",
       "    gene_id = gene_ids[[taskid]]\n",
       "    chrom = geneinfo[geneinfo$gene_id == gene_id, \"chrom\"]\n",
       "    from = geneinfo[geneinfo$gene_id == gene_id, \"start\"] - qtl_distance\n",
       "    to = geneinfo[geneinfo$gene_id == gene_id, \"start\"] + qtl_distance\n",
       "    if (file.exists(paste(\"/frazer01/projects/GTEx_v7/analysis/eqtls_deconvolution/qtls/\", \n",
       "        main, \"_original/input/phenotypes/\", gene_id, \".txt\", \n",
       "        sep = \"\")) == TRUE) {\n",
       "        traits = read_gwas_table(paste(\"/home/mdonovan/gtex_deconvolution/tables/\", \n",
       "            main, \"_gwas.txt\", sep = \"\"))\n",
       "        pv = get_ukbb_data(gene_id, main, chrom, from, to, traits, \n",
       "            \"PV\")\n",
       "        trait2coloc = lapply(traits, function(trait) {\n",
       "            trait2df(trait, pv)\n",
       "        })\n",
       "        qtl2coloc = lapply(tissue2name[tissue2name$main == main, \n",
       "            \"tissue\"], function(tissue) {\n",
       "            qtl2df(gene_id, tissue)\n",
       "        })\n",
       "        names(trait2coloc) = traits\n",
       "        names(qtl2coloc) = tissue2name[tissue2name$main == main, \n",
       "            \"tissue\"]\n",
       "        to_coloc = c(qtl2coloc, trait2coloc)\n",
       "        to_test = expand.grid(qtl = names(qtl2coloc), trait = names(trait2coloc), \n",
       "            stringsAsFactors = FALSE)\n",
       "        run_coloc_all_interactions(gene_id, to_coloc, to_test, \n",
       "            main)\n",
       "    }\n",
       "}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_by_main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "mains   = c(\"liver\", \"skin\")\n",
    "taskid = 2\n",
    "invisible(lapply(mains, function(main){dir.create(paste(\"private\", \"coloc\", main, sep = \"/\"), showWarnings = FALSE)}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00128   0.01110   0.10200   0.88400   0.00215 \n",
      "[1] \"PP abf for shared variant: 0.215%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 5.25e-05  1.23e-02  4.19e-03  9.83e-01  3.61e-04 \n",
      "[1] \"PP abf for shared variant: 0.0361%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 5.34e-05  1.23e-02  4.26e-03  9.83e-01  3.64e-04 \n",
      "[1] \"PP abf for shared variant: 0.0364%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0626    0.5420    0.0363    0.3150    0.0444 \n",
      "[1] \"PP abf for shared variant: 4.44%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00256   0.60100   0.00149   0.34900   0.04550 \n",
      "[1] \"PP abf for shared variant: 4.55%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00260   0.60000   0.00151   0.34800   0.04700 \n",
      "[1] \"PP abf for shared variant: 4.7%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0717    0.6210    0.0305    0.2640    0.0124 \n",
      "[1] \"PP abf for shared variant: 1.24%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00294   0.69000   0.00125   0.29300   0.01280 \n",
      "[1] \"PP abf for shared variant: 1.28%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00299   0.69000   0.00127   0.29300   0.01290 \n",
      "[1] \"PP abf for shared variant: 1.29%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0427    0.3700    0.0597    0.5170    0.0103 \n",
      "[1] \"PP abf for shared variant: 1.03%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00175   0.41100   0.00245   0.57500   0.00964 \n",
      "[1] \"PP abf for shared variant: 0.964%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00178   0.41100   0.00249   0.57500   0.00990 \n",
      "[1] \"PP abf for shared variant: 0.99%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0721    0.6240    0.0293    0.2540    0.0200 \n",
      "[1] \"PP abf for shared variant: 2%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00295   0.69300   0.00120   0.28200   0.02140 \n",
      "[1] \"PP abf for shared variant: 2.14%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00300   0.69200   0.00122   0.28200   0.02230 \n",
      "[1] \"PP abf for shared variant: 2.23%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0427    0.3700    0.0597    0.5170    0.0103 \n",
      "[1] \"PP abf for shared variant: 1.03%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00175   0.41100   0.00245   0.57500   0.00964 \n",
      "[1] \"PP abf for shared variant: 0.964%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00178   0.41100   0.00249   0.57500   0.00990 \n",
      "[1] \"PP abf for shared variant: 0.99%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0760    0.6580    0.0257    0.2220    0.0177 \n",
      "[1] \"PP abf for shared variant: 1.77%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00311   0.73000   0.00105   0.24700   0.01920 \n",
      "[1] \"PP abf for shared variant: 1.92%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00316   0.72900   0.00107   0.24600   0.02010 \n",
      "[1] \"PP abf for shared variant: 2.01%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0625    0.5410    0.0327    0.2830    0.0799 \n",
      "[1] \"PP abf for shared variant: 7.99%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00257   0.60200   0.00134   0.31500   0.07840 \n",
      "[1] \"PP abf for shared variant: 7.84%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00260   0.59900   0.00136   0.31400   0.08370 \n",
      "[1] \"PP abf for shared variant: 8.37%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0802    0.6950    0.0219    0.1900    0.0137 \n",
      "[1] \"PP abf for shared variant: 1.37%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 0.003280  0.771000  0.000896  0.210000  0.014600 \n",
      "[1] \"PP abf for shared variant: 1.46%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 0.003340  0.771000  0.000912  0.210000  0.014700 \n",
      "[1] \"PP abf for shared variant: 1.47%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0726    0.6290    0.0291    0.2520    0.0181 \n",
      "[1] \"PP abf for shared variant: 1.81%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00297   0.69700   0.00119   0.27900   0.01900 \n",
      "[1] \"PP abf for shared variant: 1.9%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00302   0.69700   0.00121   0.27900   0.01980 \n",
      "[1] \"PP abf for shared variant: 1.98%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0767    0.6640    0.0253    0.2200    0.0142 \n",
      "[1] \"PP abf for shared variant: 1.42%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00314   0.73700   0.00104   0.24300   0.01560 \n",
      "[1] \"PP abf for shared variant: 1.56%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00319   0.73700   0.00106   0.24400   0.01520 \n",
      "[1] \"PP abf for shared variant: 1.52%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0724    0.6270    0.0271    0.2350    0.0387 \n",
      "[1] \"PP abf for shared variant: 3.87%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00297   0.69800   0.00111   0.26100   0.03750 \n",
      "[1] \"PP abf for shared variant: 3.75%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00302   0.69600   0.00113   0.26000   0.03940 \n",
      "[1] \"PP abf for shared variant: 3.94%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0709    0.6140    0.0249    0.2160    0.0742 \n",
      "[1] \"PP abf for shared variant: 7.42%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00288   0.67700   0.00101   0.23800   0.08130 \n",
      "[1] \"PP abf for shared variant: 8.13%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00291   0.67100   0.00102   0.23600   0.08870 \n",
      "[1] \"PP abf for shared variant: 8.87%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2490    0.4760    0.0904    0.1730    0.0122 \n",
      "[1] \"PP abf for shared variant: 1.22%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0666    0.6570    0.0242    0.2380    0.0141 \n",
      "[1] \"PP abf for shared variant: 1.41%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2740    0.5240    0.0647    0.1240    0.0131 \n",
      "[1] \"PP abf for shared variant: 1.31%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0733    0.7230    0.0173    0.1700    0.0164 \n",
      "[1] \"PP abf for shared variant: 1.64%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2410    0.4610    0.0983    0.1880    0.0118 \n",
      "[1] \"PP abf for shared variant: 1.18%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0645    0.6360    0.0263    0.2590    0.0139 \n",
      "[1] \"PP abf for shared variant: 1.39%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "    0.202     0.386     0.100     0.191     0.121 \n",
      "[1] \"PP abf for shared variant: 12.1%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0491    0.4850    0.0243    0.2400    0.2020 \n",
      "[1] \"PP abf for shared variant: 20.2%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2230    0.4250    0.1170    0.2230    0.0118 \n",
      "[1] \"PP abf for shared variant: 1.18%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0596    0.5880    0.0313    0.3090    0.0127 \n",
      "[1] \"PP abf for shared variant: 1.27%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.1870    0.3580    0.1530    0.2920    0.0109 \n",
      "[1] \"PP abf for shared variant: 1.09%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0501    0.4940    0.0409    0.4030    0.0118 \n",
      "[1] \"PP abf for shared variant: 1.18%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2230    0.4250    0.1170    0.2230    0.0118 \n",
      "[1] \"PP abf for shared variant: 1.18%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0596    0.5880    0.0313    0.3090    0.0127 \n",
      "[1] \"PP abf for shared variant: 1.27%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.1870    0.3580    0.1530    0.2920    0.0109 \n",
      "[1] \"PP abf for shared variant: 1.09%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0501    0.4940    0.0409    0.4030    0.0118 \n",
      "[1] \"PP abf for shared variant: 1.18%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "    0.180     0.344     0.160     0.305     0.011 \n",
      "[1] \"PP abf for shared variant: 1.1%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0483    0.4760    0.0427    0.4210    0.0111 \n",
      "[1] \"PP abf for shared variant: 1.11%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2190    0.4180    0.1200    0.2300    0.0136 \n",
      "[1] \"PP abf for shared variant: 1.36%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0584    0.5760    0.0321    0.3170    0.0169 \n",
      "[1] \"PP abf for shared variant: 1.69%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2210    0.4220    0.1190    0.2270    0.0116 \n",
      "[1] \"PP abf for shared variant: 1.16%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0591    0.5830    0.0318    0.3130    0.0129 \n",
      "[1] \"PP abf for shared variant: 1.29%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2530    0.4840    0.0752    0.1440    0.0444 \n",
      "[1] \"PP abf for shared variant: 4.44%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0662    0.6530    0.0197    0.1940    0.0674 \n",
      "[1] \"PP abf for shared variant: 6.74%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "    0.201     0.383     0.139     0.265     0.012 \n",
      "[1] \"PP abf for shared variant: 1.2%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0536    0.5290    0.0371    0.3660    0.0144 \n",
      "[1] \"PP abf for shared variant: 1.44%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2630    0.5020    0.0764    0.1460    0.0123 \n",
      "[1] \"PP abf for shared variant: 1.23%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0701    0.6920    0.0204    0.2010    0.0169 \n",
      "[1] \"PP abf for shared variant: 1.69%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2560    0.4890    0.0754    0.1440    0.0357 \n",
      "[1] \"PP abf for shared variant: 3.57%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0662    0.6530    0.0195    0.1920    0.0696 \n",
      "[1] \"PP abf for shared variant: 6.96%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2620    0.5000    0.0779    0.1490    0.0119 \n",
      "[1] \"PP abf for shared variant: 1.19%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0699    0.6900    0.0208    0.2050    0.0145 \n",
      "[1] \"PP abf for shared variant: 1.45%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2610    0.4980    0.0789    0.1510    0.0121 \n",
      "[1] \"PP abf for shared variant: 1.21%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0696    0.6870    0.0211    0.2080    0.0149 \n",
      "[1] \"PP abf for shared variant: 1.49%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2490    0.4760    0.0895    0.1710    0.0138 \n",
      "[1] \"PP abf for shared variant: 1.38%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0666    0.6570    0.0239    0.2360    0.0172 \n",
      "[1] \"PP abf for shared variant: 1.72%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2090    0.4000    0.1280    0.2450    0.0179 \n",
      "[1] \"PP abf for shared variant: 1.79%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0559    0.5510    0.0342    0.3370    0.0213 \n",
      "[1] \"PP abf for shared variant: 2.13%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "    0.201     0.383     0.139     0.265     0.012 \n",
      "[1] \"PP abf for shared variant: 1.2%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0536    0.5290    0.0371    0.3660    0.0144 \n",
      "[1] \"PP abf for shared variant: 1.44%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2720    0.5190    0.0666    0.1270    0.0159 \n",
      "[1] \"PP abf for shared variant: 1.59%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0721    0.7110    0.0177    0.1740    0.0250 \n",
      "[1] \"PP abf for shared variant: 2.5%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2630    0.5020    0.0764    0.1460    0.0123 \n",
      "[1] \"PP abf for shared variant: 1.23%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0701    0.6920    0.0204    0.2010    0.0169 \n",
      "[1] \"PP abf for shared variant: 1.69%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.2610    0.4980    0.0781    0.1490    0.0135 \n",
      "[1] \"PP abf for shared variant: 1.35%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0696    0.6860    0.0208    0.2050    0.0180 \n",
      "[1] \"PP abf for shared variant: 1.8%\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "mains   = c(\"liver\", \"skin\")\n",
    "taskid = 2\n",
    "\n",
    "lapply(mains, function(main){coloc_by_main(main, taskid)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00128   0.01110   0.10200   0.88400   0.00215 \n",
      "[1] \"PP abf for shared variant: 0.215%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 5.25e-05  1.23e-02  4.19e-03  9.83e-01  3.61e-04 \n",
      "[1] \"PP abf for shared variant: 0.0361%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 5.34e-05  1.23e-02  4.26e-03  9.83e-01  3.64e-04 \n",
      "[1] \"PP abf for shared variant: 0.0364%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0626    0.5420    0.0363    0.3150    0.0444 \n",
      "[1] \"PP abf for shared variant: 4.44%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00256   0.60100   0.00149   0.34900   0.04550 \n",
      "[1] \"PP abf for shared variant: 4.55%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00260   0.60000   0.00151   0.34800   0.04700 \n",
      "[1] \"PP abf for shared variant: 4.7%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0717    0.6210    0.0305    0.2640    0.0124 \n",
      "[1] \"PP abf for shared variant: 1.24%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00294   0.69000   0.00125   0.29300   0.01280 \n",
      "[1] \"PP abf for shared variant: 1.28%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00299   0.69000   0.00127   0.29300   0.01290 \n",
      "[1] \"PP abf for shared variant: 1.29%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0427    0.3700    0.0597    0.5170    0.0103 \n",
      "[1] \"PP abf for shared variant: 1.03%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00175   0.41100   0.00245   0.57500   0.00964 \n",
      "[1] \"PP abf for shared variant: 0.964%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00178   0.41100   0.00249   0.57500   0.00990 \n",
      "[1] \"PP abf for shared variant: 0.99%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0721    0.6240    0.0293    0.2540    0.0200 \n",
      "[1] \"PP abf for shared variant: 2%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00295   0.69300   0.00120   0.28200   0.02140 \n",
      "[1] \"PP abf for shared variant: 2.14%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00300   0.69200   0.00122   0.28200   0.02230 \n",
      "[1] \"PP abf for shared variant: 2.23%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0427    0.3700    0.0597    0.5170    0.0103 \n",
      "[1] \"PP abf for shared variant: 1.03%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00175   0.41100   0.00245   0.57500   0.00964 \n",
      "[1] \"PP abf for shared variant: 0.964%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00178   0.41100   0.00249   0.57500   0.00990 \n",
      "[1] \"PP abf for shared variant: 0.99%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0760    0.6580    0.0257    0.2220    0.0177 \n",
      "[1] \"PP abf for shared variant: 1.77%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00311   0.73000   0.00105   0.24700   0.01920 \n",
      "[1] \"PP abf for shared variant: 1.92%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00316   0.72900   0.00107   0.24600   0.02010 \n",
      "[1] \"PP abf for shared variant: 2.01%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0625    0.5410    0.0327    0.2830    0.0799 \n",
      "[1] \"PP abf for shared variant: 7.99%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00257   0.60200   0.00134   0.31500   0.07840 \n",
      "[1] \"PP abf for shared variant: 7.84%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00260   0.59900   0.00136   0.31400   0.08370 \n",
      "[1] \"PP abf for shared variant: 8.37%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0802    0.6950    0.0219    0.1900    0.0137 \n",
      "[1] \"PP abf for shared variant: 1.37%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 0.003280  0.771000  0.000896  0.210000  0.014600 \n",
      "[1] \"PP abf for shared variant: 1.46%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 0.003340  0.771000  0.000912  0.210000  0.014700 \n",
      "[1] \"PP abf for shared variant: 1.47%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0726    0.6290    0.0291    0.2520    0.0181 \n",
      "[1] \"PP abf for shared variant: 1.81%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00297   0.69700   0.00119   0.27900   0.01900 \n",
      "[1] \"PP abf for shared variant: 1.9%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00302   0.69700   0.00121   0.27900   0.01980 \n",
      "[1] \"PP abf for shared variant: 1.98%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0767    0.6640    0.0253    0.2200    0.0142 \n",
      "[1] \"PP abf for shared variant: 1.42%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00314   0.73700   0.00104   0.24300   0.01560 \n",
      "[1] \"PP abf for shared variant: 1.56%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00319   0.73700   0.00106   0.24400   0.01520 \n",
      "[1] \"PP abf for shared variant: 1.52%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0724    0.6270    0.0271    0.2350    0.0387 \n",
      "[1] \"PP abf for shared variant: 3.87%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00297   0.69800   0.00111   0.26100   0.03750 \n",
      "[1] \"PP abf for shared variant: 3.75%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00302   0.69600   0.00113   0.26000   0.03940 \n",
      "[1] \"PP abf for shared variant: 3.94%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0709    0.6140    0.0249    0.2160    0.0742 \n",
      "[1] \"PP abf for shared variant: 7.42%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00288   0.67700   0.00101   0.23800   0.08130 \n",
      "[1] \"PP abf for shared variant: 8.13%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "  0.00291   0.67100   0.00102   0.23600   0.08870 \n",
      "[1] \"PP abf for shared variant: 8.87%\"\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OLD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_ukbb_data = function(gene_id, main, chrom, from, to, traits, type)\n",
    "{\n",
    "    outfile = paste(\"private\", \"moloc\", main, paste(gene_id, type, \"txt\", sep = \".\"), sep = \"/\")\n",
    "    command = paste(bcftools, \"query\", \n",
    "                    \"-r\", paste(gsub(\"chr\", \"\", chrom), \":\", from, \"-\", to, sep = \"\"),\n",
    "                    \"-s\", paste(traits, collapse = \",\"),\n",
    "                    \"-f\", paste('\"%POS\\t%ID\\t%AF[\\t%', type, ']\\n\"', sep = \"\"),\n",
    "                    paste(\"/frazer01/publicdata/ukbiobank_20180801/ukbb2vcf/vcf\", paste(chrom, \"vcf\", \"gz\", sep = \".\"), sep = \"/\"),\n",
    "                    \">\", \n",
    "                    outfile\n",
    "                   )\n",
    "    \n",
    "    system(command)\n",
    "    \n",
    "    #return(command)\n",
    "    \n",
    "    indata           = fread(outfile, sep = \"\\t\", header = FALSE, data.table = FALSE)\n",
    "    colnames(indata) = c(\"pos\", \"id\", \"af\", traits)\n",
    "    \n",
    "    #write.table(indata, outfile, quote = FALSE, sep = \"\\t\", row.names = FALSE, col.names = TRUE)\n",
    "    \n",
    "    return(indata)\n",
    "}\n",
    "\n",
    "bt = get_ukbb_data(gene_id, main, chrom, from, to, traits, \"BT\")\n",
    "se = get_ukbb_data(gene_id, main, chrom, from, to, traits, \"SE\")\n",
    "pv = get_ukbb_data(gene_id, main, chrom, from, to, traits, \"PV\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "qtl2df = function(gene_id, tissue)\n",
    "{\n",
    "    infile = paste(\"qtls\", tissue, \"qtls\", paste(\"all\", gene_id, \"txt\", sep = \".\"), sep = \"/\")\n",
    "    indata = fread(infile, sep = \"\\t\", header = TRUE, data.table = FALSE)\n",
    "    outdata = indata[,c(\"id\", \"beta\", \"se\", \"pval\")]\n",
    "    #colnames(outdata) = c(\"SNP\", \"BETA\", \"SE\")\n",
    "    outdata$z       = unlist(lapply(outdata$pval, function(p){zScores(p)}))\n",
    "\toutdata$varbeta = abs((outdata$beta / outdata$z)^2)\n",
    "    outdata$z       = NULL\n",
    "    \n",
    "    return(outdata)\n",
    "}\n",
    "\n",
    "trait2df = function(trait, bt, se, pv)\n",
    "{\n",
    "    this         = data.frame(id = bt$id, af = bt$af, beta = bt[,trait], se = se[,trait], pval = pv[,trait])\n",
    "    this$z       = unlist(lapply(this$pval, function(p){zScores(p)}))\n",
    "\tthis$varbeta = abs((this$beta / this$z)^2)\n",
    "    this$z       = NULL\n",
    "    \n",
    "    return(this)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "trait2coloc        = lapply(traits, function(trait){trait2df(trait, bt, se, pv)})\n",
    "qtl2coloc          = lapply(tissue2name[tissue2name$main == main, \"tissue\"], function(tissue){qtl2df(gene_id, tissue)})\n",
    "names(trait2coloc) = traits\n",
    "names(qtl2coloc)   = tissue2name[tissue2name$main == main, \"tissue\"]\n",
    "\n",
    "to_coloc = c(qtl2coloc, trait2coloc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "       NA        NA        NA        NA        NA \n",
      "[1] \"PP abf for shared variant: NA%\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 2.81e-52  7.72e-01  7.96e-53  2.19e-01  8.75e-03 \n",
      "[1] \"PP abf for shared variant: 0.875%\"\n"
     ]
    }
   ],
   "source": [
    "id1 = \"liver_cells\"\n",
    "id2 = names(trait2coloc)[[1]]\n",
    "\n",
    "b1 = to_coloc[[id1]]\n",
    "b2 = to_coloc[[id2]]\n",
    "\n",
    "rownames(b1) = b1$id\n",
    "rownames(b2) = b2$id\n",
    "\n",
    "ids = intersect(b1$id, b2$id)\n",
    "\n",
    "b1 = b1[ids,]\n",
    "b2 = b2[ids,]\n",
    "\n",
    "    \n",
    "#    b1$maf = NULL\n",
    "#    b2$maf = NULL\n",
    "#    \n",
    "#    n1 = nrow(b1)\n",
    "#    n2 = nrow(b2)\n",
    "#\n",
    "#    b1 = merge(b1, to_coloc[[gene_id]][, c(\"id\", \"maf\")])\n",
    "#    b2 = merge(b2, to_coloc[[gene_id]][, c(\"id\", \"maf\")])\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "coloc_mapped = coloc.abf(dataset1=list(snp = b1$id, pvalues = b1$pval, beta = b1$beta, varbeta = b1$varbeta, N = length(ids), MAF = b2$maf, sdY = 1, type = \"quant\"),\n",
    "                         dataset2=list(snp = b2$id, pvalues = b2$pval, beta = b2$beta, varbeta = b2$varbeta, N = length(ids), MAF = b2$maf, sdY = 1, type = \"quant\"))   \n",
    "\n",
    "coloc_mapped = coloc.abf(dataset1=list(snp = b1$id, pvalues = b1$pval, N = length(ids), type = \"quant\"),\n",
    "                         dataset2=list(snp = b2$id, pvalues = b2$pval, N = length(ids), type = \"quant\"),\n",
    "                         MAF     =b2$af\n",
    "                        )   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t7303 obs. of  5 variables:\n",
      " $ id     : chr  \"5_95211761_C_T\" \"5_95211865_A_T\" \"5_95211885_G_A\" \"5_95211992_C_T\" ...\n",
      " $ beta   : num  -0.141 -0.206 -0.206 -0.206 -0.243 ...\n",
      " $ se     : num  0.221 0.256 0.256 0.256 0.656 ...\n",
      " $ pval   : num  0.867 0.236 0.236 0.236 0.803 ...\n",
      " $ varbeta: num  0.706 0.03 0.03 0.03 0.952 ...\n",
      "'data.frame':\t7303 obs. of  6 variables:\n",
      " $ id     : chr  \"5_95211761_C_T\" \"5_95211865_A_T\" \"5_95211885_G_A\" \"5_95211992_C_T\" ...\n",
      " $ af     : num  0.48327 0.77217 0.77424 0.77218 0.00849 ...\n",
      " $ beta   : num  2.26e-05 3.92e-05 3.46e-05 3.93e-05 1.09e-04 ...\n",
      " $ se     : num  4.63e-05 5.50e-05 5.53e-05 5.50e-05 2.54e-04 ...\n",
      " $ pval   : num  0.625 0.476 0.532 0.476 0.67 ...\n",
      " $ varbeta: num  2.14e-09 3.03e-09 3.06e-09 3.03e-09 6.48e-08 ...\n"
     ]
    }
   ],
   "source": [
    "str(b1)\n",
    "str(b2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in moloc_test(tomoloc, prior_var = c(rep(0.5, length(qtl2moloc)), : Must give either MAF and N, or sdY\n",
     "output_type": "error",
     "traceback": [
      "Error in moloc_test(tomoloc, prior_var = c(rep(0.5, length(qtl2moloc)), : Must give either MAF and N, or sdY\nTraceback:\n",
      "1. moloc_test(tomoloc, prior_var = c(rep(0.5, length(qtl2moloc)), \n .     rep(0.1, length(trait2moloc))), priors = rep(1e-06, length(tomoloc)), \n .     save.SNP.info = FALSE)",
      "2. stop(\"Must give either MAF and N, or sdY\")"
     ]
    }
   ],
   "source": [
    "test = moloc_test(tomoloc, prior_var = c(rep(0.5, length(qtl2moloc)), rep(0.1, length(trait2moloc))), priors = rep(1e-6, length(tomoloc)), save.SNP.info = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 3\n",
      " $ liver_original:'data.frame':\t11768 obs. of  3 variables:\n",
      "  ..$ id  : chr [1:11768] \"5_95211761_C_T\" \"5_95211865_A_T\" \"5_95211885_G_A\" \"5_95211992_C_T\" ...\n",
      "  ..$ beta: num [1:11768] -0.0144 -0.2661 -0.2661 -0.2661 0.1761 ...\n",
      "  ..$ se  : num [1:11768] 0.231 0.265 0.265 0.265 0.689 ...\n",
      " $ liver_cells   :'data.frame':\t11768 obs. of  3 variables:\n",
      "  ..$ id  : chr [1:11768] \"5_95211761_C_T\" \"5_95211865_A_T\" \"5_95211885_G_A\" \"5_95211992_C_T\" ...\n",
      "  ..$ beta: num [1:11768] -0.141 -0.206 -0.206 -0.206 -0.243 ...\n",
      "  ..$ se  : num [1:11768] 0.221 0.256 0.256 0.256 0.656 ...\n",
      " $ liver2_cells  :'data.frame':\t11768 obs. of  3 variables:\n",
      "  ..$ id  : chr [1:11768] \"5_95211761_C_T\" \"5_95211865_A_T\" \"5_95211885_G_A\" \"5_95211992_C_T\" ...\n",
      "  ..$ beta: num [1:11768] -0.1321 -0.2122 -0.2122 -0.2122 -0.0464 ...\n",
      "  ..$ se  : num [1:11768] 0.208 0.242 0.242 0.242 0.629 ...\n"
     ]
    }
   ],
   "source": [
    "run_coloc = function(gene_id, to_coloc, id1, id2)\n",
    "{\n",
    "    b1 = to_coloc[[id1]]\n",
    "    b2 = to_coloc[[id2]]\n",
    "    \n",
    "    b1$maf = NULL\n",
    "    b2$maf = NULL\n",
    "    \n",
    "    n1 = nrow(b1)\n",
    "    n2 = nrow(b2)\n",
    "    \n",
    "    b1 = merge(b1, to_coloc[[gene_id]][, c(\"id\", \"maf\")])\n",
    "    b2 = merge(b2, to_coloc[[gene_id]][, c(\"id\", \"maf\")])\n",
    "    \n",
    "    \n",
    "    coloc_mapped = coloc.abf(dataset1=list(snp = b1$id, pvalues = b1$pval, beta = b1$beta, varbeta = b1$varbeta, N = n1, MAF = b1$maf, sdY = 1, type = \"quant\"),\n",
    "                             dataset2=list(snp = b2$id, pvalues = b2$pval, beta = b2$beta, varbeta = b2$varbeta, N = n2, MAF = b2$maf, sdY = 1, type = \"quant\"))   \n",
    "    \n",
    "    probs           = as.data.frame(t(coloc_mapped$summary))\n",
    "    myres           = coloc_mapped$results\n",
    "    myres           = coloc_mapped$results\n",
    "    myres           = myres[, c(which(colnames(myres) == \"snp\"), ncol(myres))]\n",
    "    colnames(myres) = c(\"snp_id\", \"pp_snp\")\n",
    "    myres           = myres[order(myres$pp_snp, decreasing = TRUE),]\n",
    "    \n",
    "    return(cbind(data.frame(id1 = id1, id2 = id2), probs, myres[1,]))\n",
    "}\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.5.1",
   "language": "R",
   "name": "ir35"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
